.github/workflows/release.yml

```diff
@@ -1,78 +1,130 @@
+    permissions:
+      id-token: write
+    outputs:
+      hashes: ${{ steps.hash.outputs.hashes }}
+    steps:
+      - uses: actions/checkout@d171c3b028d844f2bf14e9fdec0c58114451e4bf
+
+      - uses: actions/setup-python@7f80679172b057fc5e90d70d197929d454754a5a
+
+      - name: deps
+        run: python -m pip install -U build
+
+      - name: build
+        run: python -m build
+
+      - name: sign
+        run: |
+          mkdir -p smoketest-artifacts
+
+          # we smoke-test sigstore by installing each of the distributions
+          # we've built in a fresh environment and using each to sign and
+          # verify for itself, using the ambient OIDC identity
+          for dist in dist/*; do
+            dist_base="$(basename "${dist}")"
+
+            python -m venv smoketest-env
+
+            ./smoketest-env/bin/python -m pip install "${dist}"
+
+            # NOTE: signing artifacts currently go in a separate directory,
+            # to avoid confusing the package uploader (which otherwise tries
+            # to upload them to PyPI and fails). Future versions of twine
+            # and the gh-action-pypi-publish action should support these artifacts.
+            ./smoketest-env/bin/python -m \
+              sigstore sign "${dist}" \
+              --output-signature smoketest-artifacts/"${dist_base}.sig" \
+              --output-certificate smoketest-artifacts/"${dist_base}.crt"
+
+            ./smoketest-env/bin/python -m \
+              sigstore verify "${dist}" \
+              --cert "smoketest-artifacts/${dist_base}.crt" \
+              --signature "smoketest-artifacts/${dist_base}.sig" \
+              --cert-oidc-issuer https://token.actions.githubusercontent.com
+
+            rm -rf smoketest-env
+          done
+
+      - name: Generate hashes for provenance
+        shell: bash
+        id: hash
+        run: |
+          # sha256sum generates sha256 hash for all artifacts.
+          # base64 -w0 encodes to base64 and outputs on a single line.
+          # sha256sum artifact1 artifact2 ... | base64 -w0
+          echo "::set-output name=hashes::$(sha256sum ./dist/* | base64 -w0)"
+
+      - name: Upload built packages
+        uses: actions/upload-artifact@v3
+        with:
+          name: built-packages
+          path: ./dist/
+          if-no-files-found: warn
+
+      - name: Upload smoketest-artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: smoketest-artifacts
+          path: smoketest-artifacts/
+          if-no-files-found: warn
+
+  generate-provenance:
+    needs: [build]
+    name: Generate build provenance
+    permissions:
+      actions: read   # To read the workflow path.
+      id-token: write # To sign the provenance.
+      contents: write # To add assets to a release.
+    uses: slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@v1.2.0
+    with:
+      attestation-name: sigstore-python-provenance.intoto.jsonl
```

Comment:
This seems arbitrary. Is it possible to make this based on the name of the artifact that the provenance is generated for? Also, how does this handle multiple artifacts? Are they all bundled in the same file?

E.g., could we make the provenance for `sigstore-0.6.5.tar.gz` be `sigstore-0.6.5.tar.gz.intoto.jsonl`, and so on?

Reply:
Based on my local testings, you can use that unique provenance file to verify both artifacts. That's why I used an "arbitrary" name

Reply:
We could indeed separate the provenances, but AFAIK for that I would need to create and output two different hashes and then call the slsa-github-generator twice, which seems a bit redundant

Reply:
I'm personally okay with them being bundled together, although I also have a slight preference for the attestation filename having the version number in it somewhere (that way, someone who wants to pull down every attestation we've ever generated doesn't need to do the renaming on their own).

Reply:
Perhaps this is better as a feature request upstream for the generator. I think in the Python world, it makes more sense to have provenance be per-artifact instead.

Reply:
I'm fine with this for now if we can open an issue upstream as well.

Reply:
I understand. I'll talk to someone that I know who works on SLSA project and ask if they have this feature in mind and possibly raise the issue.
In the meanwhile, I'll try to change the code to automatically fetch the release version and create the provenance with the name `provenance-sigstore-{version number}.intoto.jsonl`. Does it sound good?

Reply:
Just commited the proposed changes, please take a look

Reply:
Hey, @laurentsimon from the upstream SLSA generator project.

The SLSA specs currently encourage the use of a single `multiple.intoto.jsonl` provenance covering all artifacts (see https://github.com/in-toto/attestation/issues/107). Having a version is definitely good feedback for them, so I replied to the thread.

> I think in the Python world, it makes more sense to have provenance be per-artifact instead.

Can you explain why it makes more sense?

Reply:
It's not the case here, but when Python builds include native code, builds have to be run across multiple platforms and architectures. In a hypothetical future, we might have a project that needs to generate provenance from multiple discrete SLSA-compatible builders, for multiple artifacts. So, the first problem is: what we name that provenance, so at verification time, users know what to use with a given artifact?

Then, if we then want to distribute provenance along with the artifact, we have the same problem: on a repository like PyPI, we have distribution files, and then additional artifacts that directly relate to those files. So for example, if I was signing something with PGP, I would upload `sigstore-0.6.5.tar.gz` and `sigstore-0.6.5.tar.gz.asc`. I think we want to follow a similar pattern here.

Reply:
Makes sense. @diogoteles08 when you get to a project like the one @di describes, let me know. Feel free to create a tracking issue on the slsa-github-generator repo as well now